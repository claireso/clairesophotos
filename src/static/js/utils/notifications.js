import urlBase64ToUint8Array from './urlBase64ToUint8Array'

export default {
  areDenied() {
    return Notification.permission === 'denied'
  },

  areGranted() {
    return Notification.permission === 'granted'
  },

  areDefault() {
    return Notification.permission === 'default'
  },

  async subscribe(registration) {
    const pushPublicKey = await this.getPushPublicKey()

    if (!registration) {
      registration = await this.getRegistration()
    }

    if (!pushPublicKey) return

    const convertedPushPublicKey = urlBase64ToUint8Array(pushPublicKey)

    const subscription = await registration.pushManager.subscribe({
      userVisibleOnly: true,
      applicationServerKey: convertedPushPublicKey
    })

    await fetch('/subscriptions', {
      method: 'post',
      headers: {
        'Content-type': 'application/json'
      },
      body: JSON.stringify({
        subscription: subscription
      })
    })
  },

  async getRegistration() {
    try {
      return await navigator.serviceWorker.ready
    } catch (err) {
      throw new Error('Banner: can not get registration')
    }
  },

  async getPushPublicKey() {
    try {
      const response = await fetch('/push-public-key')
      return await response.text()
    } catch (err) {
      throw new Error('Banner can not get public key')
    }
  }
}
