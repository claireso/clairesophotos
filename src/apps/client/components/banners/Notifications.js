import React, { useState, useEffect, useContext } from 'react'
import styled from 'styled-components'

import TranslationsContext from '@common/context/Translations'
import notifications from '@common/utils/notifications'

import Flash from '../Flash'

const ButtonSubscribe = styled.a.attrs(() => ({
  href: '#',
  role: 'button'
}))`
  color: inherit;
  text-decoration: none;

  &:before {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
  }

  &:hover {
    border-bottom: 1px solid currentColor;
  }
`

/**
 * checkSubscription: check if subscription is expired
 * @return Boolean
 */
const checkSubscription = async () => {
  try {
    const subscription = await notifications.getSubscription()

    // expired subscription
    if (!subscription) {
      return true
    }

    return false
  } catch (err) {
    throw new Error('Banner: Can not check subscription')
  }
}

export default () => {
  const [isVisible, setIsVisible] = useState(false)
  const translations = useContext(TranslationsContext)

  const subscribe = async event => {
    event && event.preventDefault()

    try {
      await notifications.subscribe()
      hideBanner()
    } catch (err) {
      // user decline
      if (notifications.areDenied()) {
        hideBanner()
        return
      }

      throw new Error('Banner: can not subscribe')
    }
  }

  const hideBanner = () => setIsVisible(false)
  const showBanner = () => setIsVisible(true)

  useEffect(() => {
    // do not display banner if service worker not support
    // do not display banner if push not enabled
    // do not display banner in safari
    if (
      !('serviceWorker' in navigator) ||
      (process.env.NODE_ENV !== 'production' &&
        process.env.NODE_ENV !== 'test') ||
      !process.env.IS_PUSH_ENABLED ||
      window.safari
    ) {
      return
    }

    // show banner if user
    // - has not already subscribed
    // - or has not denied notification
    if (notifications.areDefault()) {
      showBanner()
      return
    }

    // check if subscription is expired
    if (notifications.areGranted()) {
      checkSubscription()
        .then(isSubscriptionExpired => {
          if (isSubscriptionExpired) {
            subscribe()
          }
        })
        .catch(() => {})
    }
  }, [])

  if (!isVisible) return null

  return (
    <Flash onClose={hideBanner}>
      <ButtonSubscribe onClick={subscribe}>
        {translations.bannerNotifications}
      </ButtonSubscribe>
    </Flash>
  )
}
