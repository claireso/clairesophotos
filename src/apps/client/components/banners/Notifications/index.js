import React, { useState, useEffect, useContext, useCallback } from 'react'
import styled from 'styled-components'

import TranslationsContext from '@common/context/Translations'
import notifications from '@common/utils/notifications'

import Flash from '../../Flash'

const ButtonSubscribe = styled.a.attrs(() => ({
  href: '#',
  role: 'button'
}))`
  color: inherit;
  text-decoration: none;

  &:before {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
  }

  &:hover {
    border-bottom: 1px solid currentColor;
  }
`

export default () => {
  const [isVisible, setIsVisible] = useState(false)
  const translations = useContext(TranslationsContext)

  const subscribe = useCallback(async event => {
    event && event.preventDefault()

    try {
      await notifications.subscribe()
      hideBanner()
    } catch (err) {
      // user decline
      if (notifications.areDenied()) {
        hideBanner()
        return
      }

      throw new Error('Banner: can not subscribe')
    }
  }, [])

  const hideBanner = useCallback(() => setIsVisible(false), [])
  const showBanner = useCallback(() => setIsVisible(true), [])

  useEffect(() => {
    // do not display banner if service worker or Notification is not support
    // do not display banner if push is not enabled
    // do not display banner in safari
    if (
      !process.env.IS_PUSH_ENABLED ||
      !('serviceWorker' in navigator) ||
      !('Notification' in window) ||
      (process.env.NODE_ENV !== 'production' &&
        process.env.NODE_ENV !== 'test') ||
      window.safari
    ) {
      return
    }

    // do not display banner if if user has denied notifications
    if (notifications.areDenied()) {
      return
    }

    // check if user has already subscribed
    notifications
      .getSubscription()
      .then(subscription => {
        if (!subscription) {
          showBanner()
        }
      })
      .catch(err => {
        throw new Error('Banner can not get subscription')
      })
  }, [])

  if (!isVisible) return null

  return (
    <Flash onClose={hideBanner}>
      <ButtonSubscribe onClick={subscribe}>
        {translations.bannerNotifications}
      </ButtonSubscribe>
    </Flash>
  )
}
