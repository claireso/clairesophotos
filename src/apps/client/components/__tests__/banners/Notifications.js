import React from 'react'
import { render, fireEvent } from 'react-testing-library'

import TranslationsContext from '@common/context/Translations'
import notifications from '@common/utils/notifications'

import Notifications from '../../banners/Notifications'

describe('<Notifications />', () => {
  test('should render component', () => {
    setNotificationPermission()

    const { container } = render(
      <TranslationsContext.Provider value={__TRANSLATIONS__.client}>
        <Notifications />
      </TranslationsContext.Provider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should not render component', () => {
    setNotificationPermission('denied')

    const { container } = render(
      <TranslationsContext.Provider value={__TRANSLATIONS__.client}>
        <Notifications />
      </TranslationsContext.Provider>
    )

    expect(container).toMatchSnapshot()
    setNotificationPermission()
  })

  test('should subscribe and hide banner', async () => {
    setNotificationPermission()

    // mock func subscribe
    const spy = jest
      .spyOn(notifications, 'subscribe')
      .mockImplementation(() => Promise.resolve())

    const { container, getByRole } = render(
      <TranslationsContext.Provider value={__TRANSLATIONS__.client}>
        <Notifications />
      </TranslationsContext.Provider>
    )

    await fireEvent.click(getByRole('button'))

    expect(container).toMatchSnapshot()
    spy.mockRestore()
  })

  test('should not subscribe and hide banner (denied story)', async () => {
    setNotificationPermission()

    // mock func subscribe
    const spy = jest
      .spyOn(notifications, 'subscribe')
      .mockImplementation(() => Promise.reject())

    const { container, getByRole, debug } = render(
      <TranslationsContext.Provider value={__TRANSLATIONS__.client}>
        <Notifications />
      </TranslationsContext.Provider>
    )

    setNotificationPermission('denied')

    await fireEvent.click(getByRole('button'))

    expect(container).toMatchSnapshot()
    spy.mockRestore()
    setNotificationPermission()
  })

  test('should not render component and not call `subscribe` function (registration not expired)', done => {
    setNotificationPermission('granted')

    const spy = jest
      .spyOn(notifications, 'getSubscription')
      .mockImplementation(() => Promise.resolve({}))
    const spySubscribe = jest
      .spyOn(notifications, 'subscribe')
      .mockImplementation(() => Promise.resolve())

    const { container } = render(
      <TranslationsContext.Provider value={__TRANSLATIONS__.client}>
        <Notifications />
      </TranslationsContext.Provider>
    )

    setImmediate(() => {
      expect(spySubscribe).toHaveBeenCalledTimes(0)
      expect(container).toMatchSnapshot()

      spy.mockRestore()
      spySubscribe.mockRestore()
      done()
    })
  })

  test('should not render component and call `subscribe` function (registration expired)', done => {
    setNotificationPermission('granted')

    const spy = jest
      .spyOn(notifications, 'getSubscription')
      .mockImplementation(() => Promise.resolve())
    const spySubscribe = jest
      .spyOn(notifications, 'subscribe')
      .mockImplementation(() => Promise.resolve())

    const { container } = render(
      <TranslationsContext.Provider value={__TRANSLATIONS__.client}>
        <Notifications />
      </TranslationsContext.Provider>
    )

    setImmediate(() => {
      expect(spySubscribe).toHaveBeenCalledTimes(1)
      expect(container).toMatchSnapshot()
      spy.mockRestore()
      spySubscribe.mockRestore()
      done()
    })
  })
})
