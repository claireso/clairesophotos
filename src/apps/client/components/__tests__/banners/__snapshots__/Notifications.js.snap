// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifications /> should not render component 1`] = `<div />`;

exports[`<Notifications /> should not render component and call \`subscribe\` function (registration expired) 1`] = `<div />`;

exports[`<Notifications /> should not render component and not call \`subscribe\` function (registration not expired) 1`] = `<div />`;

exports[`<Notifications /> should not subscribe and hide banner (denied story) 1`] = `<div />`;

exports[`<Notifications /> should render component 1`] = `
.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  cursor: pointer;
  outline: none;
  padding: 1rem 1.2rem 1.1rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 {
  background: var(--yellow);
  color: var(--white);
  font-size: 1.4rem;
  margin: 0 0 var(--gutter);
  padding: 1.5rem;
  position: relative;
  text-align: center;
}

.c1 button {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
}

.c1 + .c0 {
  border-top: 3px solid var(--yellow-darker);
}

.c2 {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.c2:hover {
  border-bottom: 1px solid currentColor;
}

<div>
  <div
    class="c0 c1"
    style="opacity: 0;"
  >
    <a
      class="c2"
      href="#"
      role="button"
    >
      Activer les notifications pour être alerté de nouvelle publication
    </a>
    <button
      class="c3"
    >
      <svg
        height="16"
        viewBox="0 0 22 28"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.281 20.656c0 0.391-0.156 0.781-0.438 1.062l-2.125 2.125c-0.281 0.281-0.672 0.438-1.062 0.438s-0.781-0.156-1.062-0.438l-4.594-4.594-4.594 4.594c-0.281 0.281-0.672 0.438-1.062 0.438s-0.781-0.156-1.062-0.438l-2.125-2.125c-0.281-0.281-0.438-0.672-0.438-1.062s0.156-0.781 0.438-1.062l4.594-4.594-4.594-4.594c-0.281-0.281-0.438-0.672-0.438-1.062s0.156-0.781 0.438-1.062l2.125-2.125c0.281-0.281 0.672-0.438 1.062-0.438s0.781 0.156 1.062 0.438l4.594 4.594 4.594-4.594c0.281-0.281 0.672-0.438 1.062-0.438s0.781 0.156 1.062 0.438l2.125 2.125c0.281 0.281 0.438 0.672 0.438 1.062s-0.156 0.781-0.438 1.062l-4.594 4.594 4.594 4.594c0.281 0.281 0.438 0.672 0.438 1.062z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`<Notifications /> should subscribe and hide banner 1`] = `<div />`;
